packages/SwingSet/misc-tools/audit-refgraph.py

```diff
@@ -0,0 +1,115 @@
+        return reachable_vrefs
+
+    def audit(self):
+        assert(int(self.meta["|entryCount"]) == len(self.data))
+        
+    def audit_ordinals(self):
+        for c_key in self.data:
+            if c_key.startswith("r"):
+                # data record, where the key is an ordinal
+                # 'r0000000003:o+d31/35:1' -> '{"body":"#null","slots":[]}'
+                pieces = c_key.split(":", maxsplit=1)
+                dr_ordinal_s = pieces[0][1:]
+                dr_ordinal = int(dr_ordinal_s)
+                dr_ordinal_vref = pieces[1]
+                if dr_ordinal_vref not in self.ordinals:
+                    raise ValueError("%s.c%s vref=%s ordinal=%s" % (self.vatID, self.collectionID, dr_ordinal_vref, dr_ordinal))
+                oa_ordinal = self.ordinals[dr_ordinal_vref]
+                if dr_ordinal != oa_ordinal:
+                    raise ValueError("%s.c%s vref=%s ordinal=%s/%s" % (self.vatID, self.collectionID, dr_ordinal_vref, dr_ordinal, oa_ordinal))
+        for oa_vref, oa_ordinal in self.ordinals.items():
+            dr_key = "r%010d:%s" % (oa_ordinal, oa_vref)
+            if dr_key not in self.data:
+                raise ValueError("%s.c%s vref=%s ordinal=%s" % (self.vatID, self.collectionID, oa_vref, oa_ordinal))
+
+
+def make_vatID():
+    return defaultdict(Collection)
+
+class Vat:
+    def __init__(self, vatID):
+        self.vatID = vatID
+        self.collections = {}
+        self.refcounts = {} # vom.rc.$vref: $vref -> reachable_count
+        self.refs = defaultdict(set) # $vref -> inbound vrefs
+
+    def add_line(self, v_line, value):
+        if v_line.startswith("vs."):
+            vs_key = v_line[len("vs."):]
+            self.add_vatstore(vs_key, value)
+    def add_vatstore(self, vs_key, value): # vom.rc. or vc.$cid. , etc
+        #print(vatID, vs_key, value)
+        mo = vcRE.match(vs_key) # vc.$cid.
+        if mo:
+            (collectionID, c_key) = mo.groups()
+            collectionID = int(collectionID)
+            if collectionID not in self.collections:
+                self.collections[collectionID] = Collection(self.vatID, collectionID)
+            c = self.collections[collectionID]
+            reachable_vrefs = c.add(c_key, value)
+            for vref in reachable_vrefs:
+                self.refs[vref].add("collection-%d" % collectionID)
+        mo = rcRE.search(vs_key) # vom.rc.
+        if mo:
+            vref = mo.group(1)
+            reachable_s = value
+            self.refcounts[vref] = int(reachable_s)
+        
+        
+
+collections_by_vatID = {}
+
+
+for row in sys.stdin:
+    key,value = row.strip().split("_", maxsplit=1)
+    mo = vatRE.search(key)
+    if mo:
+        (vatID, rest)  = mo.groups()
+        if vatID not in collections_by_vatID:
+            collections_by_vatID[vatID] = Vat(vatID)
+        v = collections_by_vatID[vatID]
+        v.add_line(rest, value)
+
+c = collections_by_vatID["v9"].collections[94]
```

Comment:
What does `94` identify?

Reply:
oh, that was leftover from experimenting/investigation I was doing. I'll mark it as such. (many of these tools are.. interactive artworks, and are configured by modifying the source rather than by taking arguments).